topdown/http.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2018 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package topdown
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+)
+
+// DefaultHTTPRequestTimeoutSec Default timeout value for http client
+const defaultHTTPRequestTimeoutSec = 5
+
+var allowedKeys = ast.NewSet(ast.StringTerm("method"), ast.StringTerm("url"), ast.StringTerm("body"))
+var requiredKeys = ast.NewSet(ast.StringTerm("method"), ast.StringTerm("url"))
+
+var client *http.Client
+
+func builtinHTTPReq(bctx BuiltinContext, a ast.Value) (ast.Value, error) {
+	switch a.(type) {
+	case ast.Object:
+		obj, ok := a.(ast.Object)
+		if !ok {
+			return nil, fmt.Errorf("type assertion error")
+		}
+
+		return evaluateHTTPRequest(obj, bctx)
+	default:
+		return nil, builtins.NewOperandTypeErr(1, a, ast.ObjectTypeName)
+	}
+
+}
+
+func init() {
+	createHTTPClient()
+	RegisterFunctionalBuiltin1WithCtxt(ast.HTTPReqBuiltin.Name, builtinHTTPReq)
+}
+
+// createHTTPClient creates a HTTP client
+func createHTTPClient() {
+	var timeout time.Duration
+
+	timeoutDuration := os.Getenv("HTTP_SEND_TIMEOUT")
+
+	if timeoutDuration == "" {
+		timeout = time.Duration(defaultHTTPRequestTimeoutSec) * time.Second
+	} else {
+		timeout, _ = time.ParseDuration(timeoutDuration)
+	}
+
+	client = &http.Client{
+		Timeout: timeout,
+	}
+}
+
+// evaluateHTTPRequest executes the HTTP request and processes the response
+func evaluateHTTPRequest(obj ast.Object, bctx BuiltinContext) (ast.Value, error) {
+	var url string
+	var method string
+	var body *bytes.Buffer
+
+	requestKeys := ast.NewSet(obj.Keys()...)
+
+	// check invalid keys
+	invalidKeys := requestKeys.Diff(allowedKeys)
+	if invalidKeys.Len() != 0 {
+		return nil, fmt.Errorf("Invalid Key %s", invalidKeys)
+	}
+
+	// check missing keys
+	missingKeys := requiredKeys.Diff(requestKeys)
+	if missingKeys.Len() != 0 {
+		return nil, fmt.Errorf("Missing Keys %s", missingKeys)
+	}
+
+	for _, val := range obj.Keys() {
+		key := strings.Trim(val.String(), "\"")
```

Comment:
Instead of using the .String() function here, I would rather use a type assertion to convert from ast.Value -> ast.String -> string. That way you aren't dependant on the String() representation that requires trim calls.

Reply:
done